cli/tfsource/types.go

```diff
@@ -47,8 +48,31 @@ func (src *TerraformSource) String() string {
 // See also the encodeSourceName and ProcessTerraformSource methods.
 func (terraformSource TerraformSource) EncodeSourceVersion() string {
 	if IsLocalSource(terraformSource.CanonicalSourceURL) {
-		hash, err := dirhash.HashDir(terraformSource.CanonicalSourceURL.Path, "prefix", dirhash.DefaultHash)
+		sourceHash := sha256.New()
+		sourceDir := filepath.Clean(terraformSource.CanonicalSourceURL.Path)
+
+		err := filepath.Walk(sourceDir, func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				// If we've encountered an error while walking the tree, give up
+				return err
+			}
+
+			if info.IsDir() {
+				// We don't use any info from directories to calculate our hash
+				return nil
+			}
+
+			fileModified := info.ModTime().UnixMicro()
+			hashContents := fmt.Sprintf("%s:%d", path, fileModified)
+
+			sourceHash.Write([]byte(hashContents))
+
+			return nil
+		})
+
 		if err == nil {
```

Comment:
I think will be helpful to log error, this will help to debug any possible access error

Reply:
Based on #2171, yeah, it would be helpful to log a warning when there's an error hashing the directory contents...

Reply:
@denis256 Done!

Reply:
Right, I've fixed the logging code, so it should be good to go in its current form.

I do agree with you about bypassing the cache if we run into an error while validating it though. I'm going to see how tricky that would be to introduce.

Reply:
Well you could just hash a random string, yeah? :p That would make it think it changed, and force it to update?

Reply:
Okay, I've had a look around the codebase to see how this function gets used. We could probably bypass the cache if `EncodeSourceVersion` were to return an error, but it's not totally straightforward at all the callsites.

  - [This case](https://github.com/gruntwork-io/terragrunt/blob/7d143740c05ef9a5cc87414c5b22107bf3cb1425/cli/download_source.go#L112) and [this case](https://github.com/gruntwork-io/terragrunt/blob/7d143740c05ef9a5cc87414c5b22107bf3cb1425/cli/download_source.go#L147) seem reasonably easy to handle. We could have them assume the download folder didn't have an up-to-date copy of the source and copy it over again.
  - [This case](https://github.com/gruntwork-io/terragrunt/blob/7d143740c05ef9a5cc87414c5b22107bf3cb1425/cli/tfsource/types.go#L62) is slightly more painful. Returning an error would cause Terragrunt to exit without doing anything as the error would be propagated all the way back to `cli_app.go`. We could do something weird like write a random value into the file there, effectively signalling "we don't know what the hash was, so you're gonna have to copy this module again next time", but it feels somewhat janky.

The other thing that would need taking care of is some testing. I haven't looked at the test suite/testing libraries in use to see if we'd have an easy way of stubbing `EncodeSourceVersion` to return an error. I'm sure it's doable, but I know stubbing/mocking can be a bit of a pain in Go if you don't have an interface to act as a testing seam.

Reply:
Ok, I think it should work fine with log + random hash

Reply:
Also, if on error it's going to force-update the cache, which requires less from the user to "just work", then I think logging at the debug level is fine. I suggested a warning previously just because of what we were seeing in #2171, where a hash error may require user interaction to clean up their working dir, and so the visibility/noise of a warning was worth it.

Reply:
Cool - I'll add that in.

I've got the day off tomorrow, so I'm not sure I'll look at this until Monday, but I'll add an extra commit to this PR for the new hashing behaviour then.

Reply:
Cool - I'll add that in.

I've got the day off tomorrow, so I'm not sure I'll look at this until Monday, but I'll add an extra commit to this PR for the new hashing behaviour then.

Reply:
@denis256 Alright, that's in there now as a second commit!

Testing the error case was a little tricky. Best I could come up with was making the containing directory have `0000` permissions so that the `stat` call failed and put us into our error handling code.